sale_partner_delivery_window/models/sale_order.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+import logging
+from datetime import timedelta
+
+from odoo import _, api, fields, models
+from odoo.tools.misc import format_datetime
+
+_logger = logging.getLogger(__name__)
+
+
+class SaleOrder(models.Model):
+
+    _inherit = "sale.order"
+
+    @api.depends(
+        "partner_shipping_id.delivery_time_preference",
+        "partner_shipping_id.delivery_time_window_ids",
+        "partner_shipping_id.delivery_time_window_ids.time_window_start",
+        "partner_shipping_id.delivery_time_window_ids.time_window_end",
+        "partner_shipping_id.delivery_time_window_ids.time_window_weekday_ids",
+    )
+    def _compute_expected_date(self):
+        """Add dependencies to consider fixed delivery windows"""
+        return super()._compute_expected_date()
+
+    @api.onchange("commitment_date")
+    def _onchange_commitment_date(self):
+        """Warns if commitment date is not a preferred window for delivery"""
+        res = super()._onchange_commitment_date()
+        if res:
+            return res
+        if (
+            self.commitment_date
+            and self.partner_shipping_id.delivery_time_preference == "time_windows"
+        ):
+            ps = self.partner_shipping_id
+            if not ps.is_in_delivery_window(self.commitment_date):
+                return {
+                    "warning": {
+                        "title": _(
+                            "Commitment date does not match shipping "
+                            "partner's Delivery time schedule preference."
+                        ),
+                        "message": _(
+                            "The delivery date is %s, but the shipping "
+                            "partner is set to prefer deliveries on following "
+                            "time windows:\n%s"
+                            % (
+                                format_datetime(self.env, self.commitment_date),
+                                "\n".join(
+                                    [
+                                        "  * %s" % w.display_name
+                                        for w in ps.get_delivery_windows().get(ps.id)
+                                    ]
+                                ),
+                            )
+                        ),
+                    }
+                }
+
+
+class SaleOrderLine(models.Model):
+
+    _inherit = "sale.order.line"
+
+    def _expected_date(self):
+        """Postpone expected_date to next preferred delivery window"""
+        expected_date = super()._expected_date()
+        partner = self.order_id.partner_shipping_id
+        if partner.delivery_time_preference == "anytime":
+            return expected_date
+        return partner.next_delivery_window_start_datetime(from_date=expected_date)
```

Comment:
Tested locally: if the partner is still not defined on the SO while we enter SO lines, the following error is raised:
```python
[...]
  File "/odoo/src/addons/sale_stock/models/sale_order.py", line 53, in _compute_expected_date
    super(SaleOrder, self)._compute_expected_date()
  File "/odoo/src/addons/sale/models/sale.py", line 294, in _compute_expected_date
    dt = line._expected_date()
  File "/odoo/external-src/sale-workflow/sale_partner_delivery_window/models/sale_order.py", line 73, in _expected_date
    return partner.next_delivery_window_start_datetime(from_date=expected_date)
  File "/odoo/external-src/sale-workflow/sale_partner_delivery_window/models/res_partner.py", line 23, in next_delivery_window_start_datetime
    self.ensure_one()
  File "/odoo/src/odoo/models.py", line 5003, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: res.partner()
```